### DISCUSSION TITLE
Discussion on PyStackAPI project.

### DISCUSSION THREAD

**Discussion by wchistow** [wchistow/pystackapi#2] [Announcements]
Write something here.

**Comment by Peter-developer01** [comment]
This is cool project. Maybe you want add me to a contributors, I know Python.

This can be great experience for you and me.

**Reply by wchistow** [reply]
I've added you to contributors. As a trial, implement the `Site.get_answers` method  like `get_questions`. PS: API method documentation: https://api.stackexchange.com/docs/answers

**Reply by Peter-developer01** [reply]
@wchistow for info, you should do "git pull" when you start working on project and probably every hour. Just for info.

**Reply by wchistow** [reply]
@Peter-developer01, warning: I've added the code checking with MyPy on push - https://github.com/wchistow/pystackapi/actions/workflows/mypy.yml.

**Reply by wchistow** [reply]
@Peter-developer01, note: I've added auto running tests on push. So, before each commit, run mypy and tests. MyPy runs by command `python -m mypy ./src/ --config-file ./mypy.ini`, requirements is `mypy` and `types-requests`. About tests, they run by `cd tests; python main.py`, and the requirements may be installed by `python -m pip install rich==13.4.2; python -m pip install -r ./requirements.txt`.

**Reply by Peter-developer01** [reply]
@wchistow before each *commit* or before each *push*?

**Reply by wchistow** [reply]
@Peter-developer01, better before committing, so as not to commit non-working code.

**Reply by Peter-developer01** [reply]
@wchistow ok.

**Comment by wchistow** [comment]
I think, that this code from example getting general information of site is not very beautiful:

```py
response = site.get_info()

information = response['items'][0]
#                     ^^^^^^^^^^^^
```

My idea is that `get_info` and `get_<singular_noun>` returns `Item`, and, accordingly, `get_<plural_noun>` returns `list[Item]` .  Remind you that the API response contains the following keys:

 + `items`
 + `has_more`
 + `quota_max`
 + `quota_remaining`

So, in my solution these keys, except `items`, are missing.

But maybe they are not so important?

So, I want to know your opinion, @Peter-developer01, is it worth doing this.

**Reply by wchistow** [reply]
I've come up with a way out! `Site.get` returns raw dict, but `Site.get_<singular_noun>` returns `Item` and `get_<plural_noun>` returns `list[Item]`.

**Comment by nchistov** [comment]
Please add contributors guide file.

**Reply by wchistow** [reply]
Added - https://github.com/wchistow/pystackapi/commit/f88af4a6dea68c7fcdf5aec048f2ffb3b56c56ab

**Comment by Peter-developer01** [comment]
@wchistow sorry, I'm a little busy now.

**Reply by wchistow** [reply]
How long will you be busy?

**Reply by Peter-developer01** [reply]
I sometimes can freely contribute, but, if you want to know the exact date... this is approximately and and a half year.

**BUT**: I sometimes are free, so I can sometimes free contribute.

**Reply by wchistow** [reply]
OK

**Comment by wchistow** [comment]
@nchistov, can you delete your [fork](https://github.com/nchistov/pystackapi) of this project? In my opinion, this fork is not necessary now.

**Reply by nchistov** [reply]
Yes, I delete it.

**Comment by wchistow** [comment]
@nchistov @Peter-developer01 Hooray, PyStackAPI 0.1.0 is here!

**Reply by Peter-developer01** [reply]
Good! :tada:

**Reply by wchistow** [reply]
See question about it on StackApps: https://stackapps.com/q/10408/116701.

**Reply by Peter-developer01** [reply]
Good, I've suggested an edit.

**Comment by wchistow** [comment]
@nchistov, @Peter-developer01, we are starting the development of version 0.2.0! Join!

**Reply by Peter-developer01** [reply]
Good news! I got some more free time. So, how do I start? What should I do?

**Reply by wchistow** [reply]
@Peter-developer01, most likely, today, I'll add necessary tasks to ToDo. Now I'm trying to figure out how to send POST-requests to the StackExchange API.

**Reply by wchistow** [reply]
@Peter-developer01, I found, how to get an access token, it's so hard! About it see my question (https://ru.stackoverflow.com/q/1548376) and comment by user `strawdog` as the answer. *(PS: access token is needable for testing new methods)*

**Reply by wchistow** [reply]
@Peter-developer01, I added (0dfb10d) some tasks to ToDo. We'll start this version with GET-methods with authentication (mostly, `me/*`). *PS: don't write tests yet, they need to be redone and I'll do this tomorrow.*

**Reply by wchistow** [reply]
@Peter-developer01, like the example of this methods, see method [`get_me`](https://github.com/wchistow/pystackapi/blob/master/src/pystackapi/site.py#L353).

**Reply by Peter-developer01** [reply]
@wchistow oh heck yeah, getting access token is hard.

**Reply by Peter-developer01** [reply]
I'll try to implement methods `get_my_*`. Do you agree to test it? I don't want to get an access token.

**Reply by wchistow** [reply]
@Peter-developer01, ok.

**Reply by wchistow** [reply]
@Peter-developer01 ,by the way, you can write tests in folder [`tests/test_site/`](https://github.com/wchistow/pystackapi/tree/master/tests/test_site) yourself. As the example, see [tests for method `get_me`](https://github.com/wchistow/pystackapi/blob/master/tests/test_site/test_get_users.py#L135).

**Reply by Peter-developer01** [reply]
@wchistow How to run tests?

**Reply by wchistow** [reply]
@Peter-developer01, see [contributors' guide](https://github.com/wchistow/pystackapi/blob/master/CONTRIBUTING.md) (`cd tests; python main.py`).

**Reply by wchistow** [reply]
@Peter-developer01, about access token: I've wrote [an answer](https://ru.stackoverflow.com/a/1548871/507426) on RuSO about getting it. I wrote it for two days!

**Reply by Peter-developer01** [reply]
@wchistow Cool!

**Reply by wchistow** [reply]
@Peter-developer01, I've added (https://github.com/wchistow/pystackapi/commit/aefbd71cb8fd11e090bde1f071e67bea70a70454) some tasks to the ToDo, you can to do them.

**Reply by wchistow** [reply]
@Peter-developer01, as the example of the POST-method you can use the [`add_answer`](https://github.com/wchistow/pystackapi/blob/master/src/pystackapi/site.py#L48) method.

**Comment by Peter-developer01** [comment]
@wchistow just ignore that. No need to ask what is this.

**Comment by wchistow** [comment]
@Peter-developer01, when you add some method to the library, don't forget to add it to the [`CHANGELOG.md`](https://github.com/wchistow/pystackapi/blob/master/CHANGELOG.md) file.

**Reply by Peter-developer01** [reply]
Okay.

**Comment by wchistow** [comment]
All ToDo tasks have done, and, if you don't have any more ideas for it, tomorrow the 0.2.0 version will be out.

**Comment by wchistow** [comment]
@Peter-developer01, @nchistov, I'm going to add StackExchange chat API to the next version of PyStackAPI (0.3.0). Information about its interface is quite difficult to find (there is not a word about it in the official documentation!), but I found this file: https://github.com/Manishearth/ChatExchange/blob/master/docs/se-chat-interface.md. We'll start soon)

**Reply by Peter-developer01** [reply]
Good luck! ;)
How will you test it?

**Reply by wchistow** [reply]
@Peter-developer01, like everything else. What's the problem?

**Reply by Peter-developer01** [reply]
Okay, nevermind.

**Reply by wchistow** [reply]
@Peter-developer01, I think it would be easier to just make a wrapper for the [finished library](https://github.com/Manishearth/ChatExchange) 😉

**Reply by Peter-developer01** [reply]
Yeah it is a lot easier. If you test it yourself, you can find what you don't like and rewrite it to be easier (I mean, wrapper it) ;)

**Reply by wchistow** [reply]
*rewrite it* - And first I'll [fix](https://github.com/Manishearth/ChatExchange/pull/183) `DeprecationWarning` ;)

**Reply by Peter-developer01** [reply]
Feel free to mess with the library and see what looks better and natural. ;)

**Reply by Peter-developer01** [reply]
I have an idea... How about extend the original ChatExchange with functions to read transcript, messages, etc.? It would be good, as now I couldn't find such a feature, or at least it is not obvious how to use it.

**Comment by wchistow** [comment]
@Peter-developer01, @nchistov, today is the PyStackAPI's birthday! The first commit was created on May 28 2023, a year ago!

**Reply by Peter-developer01** [reply]
Yay!
